### PR TITLE
shader_decode: Reimplement BFE instructions

### PR DESCRIPTION
**Pull request by namkazt**
This PR reimplement BFE instruction set. Still missing ~~BREV and~~ Rd.CC tho.
As always thank to @ReinUsesLisp

### CODE REVIEW

File: src/video_core/shader/decode/bfe.cpp [hunk 1]
```diff
@@ -17,33 +17,57 @@ u32 ShaderIR::DecodeBfe(NodeBlock& bb, u32 pc) {
...
-        UNIMPLEMENTED_MSG("Unhandled BFE instruction: {}", opcode->get().GetName());
+    // using reverse parallel method in
+    // https://graphics.stanford.edu/~seander/bithacks.html#ReverseParallel
+    // note for later if possible to implement faster method.
+    if (instr.bfe.brev) {
+        const auto swap = [&](u32 s, u32 mask) {
+            Node v1 =
+                SignedOperation(OperationCode::ILogicalShiftRight, is_signed, op_a, Immediate(s));
+            if (mask != 0) {
+                v1 = SignedOperation(OperationCode::IBitwiseAnd, is_signed, v1, Immediate(mask));
```
Comment (lioncash) [inline src/video_core/shader/decode/bfe.cpp]:
```suggestion
                v1 = SignedOperation(OperationCode::IBitwiseAnd, is_signed, std::move(v1), Immediate(mask));
```

File: src/video_core/shader/decode/bfe.cpp [hunk 2]
```diff
@@ -17,33 +17,57 @@ u32 ShaderIR::DecodeBfe(NodeBlock& bb, u32 pc) {
...
+            }
+            Node v2 = op_a;
+            if (mask != 0) {
+                v2 = SignedOperation(OperationCode::IBitwiseAnd, is_signed, op_a, Immediate(mask));
```
Comment (lioncash) [inline src/video_core/shader/decode/bfe.cpp]:
```suggestion
                v2 = SignedOperation(OperationCode::IBitwiseAnd, is_signed, std::move(v2), Immediate(mask));
```

File: src/video_core/shader/decode/bfe.cpp [hunk 3]
```diff
@@ -17,33 +17,57 @@ u32 ShaderIR::DecodeBfe(NodeBlock& bb, u32 pc) {
...
+            }
+            v2 = SignedOperation(OperationCode::ILogicalShiftLeft, is_signed, v2, Immediate(s));
```
Comment (lioncash) [inline src/video_core/shader/decode/bfe.cpp]:
```suggestion
            v2 = SignedOperation(OperationCode::ILogicalShiftLeft, is_signed, std::move(v2), Immediate(s));
```

File: src/video_core/shader/decode/bfe.cpp [hunk 4]
```diff
@@ -17,33 +17,57 @@ u32 ShaderIR::DecodeBfe(NodeBlock& bb, u32 pc) {
...
+            return SignedOperation(OperationCode::IBitwiseOr, is_signed, v1, v2);
```
Comment (lioncash) [inline src/video_core/shader/decode/bfe.cpp]:
```suggestion
            return SignedOperation(OperationCode::IBitwiseOr, is_signed, std::move(v1), std::move(v2));
```

File: src/video_core/shader/decode/bfe.cpp [hunk 5]
```diff
@@ -17,33 +17,57 @@ u32 ShaderIR::DecodeBfe(NodeBlock& bb, u32 pc) {
...
+        op_a = swap(16, 0);
     }
 
+    const auto offset = SignedOperation(OperationCode::IBitfieldExtract, is_signed, op_b,
+                                        Immediate(0), Immediate(8));
+    const auto bits = SignedOperation(OperationCode::IBitfieldExtract, is_signed, op_b,
+                                      Immediate(8), Immediate(8));
+    const auto result =
+        SignedOperation(OperationCode::IBitfieldExtract, is_signed, op_a, offset, bits);
+    SetRegister(bb, instr.gpr0, result);
```
Comment (lioncash) [inline src/video_core/shader/decode/bfe.cpp]:
```suggestion
    SetRegister(bb, instr.gpr0, std::move(result));
```

**Comment by namkazt** [comment]
@lioncash I got a warning:
2020-03-13 15_48_37-Window

**Comment by ReinUsesLisp** [comment]
Remove `const`.

**Pull request merged**